src/lib/simulator/io/ParticleLoader.cpp

```diff
@@ -0,0 +1,397 @@
+      return std::nullopt;
+    }
+  }
+  if (std::isnan(mass)) {
+    spdlog::error("Invalid planet data: NaN mass encountered.");
+    return std::nullopt;
+  }
+
+  const auto p = Particle(position, velocity, mass, 0);
+  recognize_end_of_line(buf);
+  spdlog::debug("Recognized planet: position = ({}, {}, {}), velocity = ({}, {}, {}), mass = {}",
+                position[0], position[1], position[2],
+                velocity[0], velocity[1], velocity[2], mass);
+  return {p};
+}
+
+/**
+     * @brief Parses particles affected by gravity from the input buffer.
+     *
+     * @param buf Input buffer iterator.
+     * @return std::optional<std::vector<Particle>> The parsed particles or an empty optional if parsing fails.
+     */
+auto ParticleLoader::parse_gravity(std::istreambuf_iterator<char> &buf) -> std::optional<std::vector<Particle>> {
+  spdlog::trace("Starting to parse gravity particles");
+  auto particles = std::vector<Particle>();
+  while (true) {
+    if (recognize_end_of_file(buf)) {
+      spdlog::trace("Reached end of file while parsing gravity particles");
+      break;
+    }
+    if (const auto particle = recognize_planet(buf)) {
+      particles.push_back(*particle);
+    } else {
+      spdlog::warn("Failed to recognize a planet while parsing gravity particles");
+      break;
+    }
+  }
+  spdlog::debug("Parsed {} gravity particles", particles.size());
+  return particles;
+}
+
+/**
+     * @brief Parses cuboid structures from the input buffer.
+     *
+     * @param buf Input buffer iterator.
+     * @return std::optional<std::vector<cuboid_t>> The parsed cuboids or an empty optional if parsing fails.
+     */
+auto ParticleLoader::parse_cuboids(std::istreambuf_iterator<char> &buf) -> std::optional<std::vector<cuboid_t>> {
+  spdlog::trace("Starting to parse cuboids");
+  auto particles = std::vector<cuboid_t>();
+  while (true) {
+    if (recognize_end_of_file(buf)) {
+      spdlog::trace("Reached end of file while parsing cuboids");
+      break;
+    }
+    if (const auto particle = recognize_cuboid(buf)) {
+      particles.push_back(*particle);
+    } else {
+      spdlog::warn("Failed to recognize a cuboid while parsing");
+      break;
+    }
+  }
+  spdlog::debug("Parsed {} cuboids", particles.size());
+  return particles;
+}
+
+// TODO make configurable
+static constexpr const double brownian_motion = 0.1;
+
+/**
+    * @brief Generates particles from a list of cuboids using the given seed.
+    *
+    * Iterates through each cuboid and generates particles based on its dimensions and properties.
+    * Logs the progress at various levels.
+    *
+    * @param cuboids A vector of cuboid_t structures containing cuboid properties.
+    * @param seed The seed used for random number generation.
+    * @return std::vector<Particle> The generated particles.
+    */
+auto ParticleLoader::generate_cuboids(const std::vector<cuboid_t> &cuboids, auto seed) -> std::vector<Particle> {
```

Comment:
Is there a reason you went for `auto function() -> return_type` format? I would stick to the standard way, `return_type function()`, for better readability unless there is a strong reason not to